pkg/chunkenc/symbols.go

```diff
@@ -0,0 +1,374 @@
+		return 0, nil, errors.Wrap(err, "appending entry")
+	}
+	if err := compressedWriter.Close(); err != nil {
+		return 0, nil, errors.Wrap(err, "flushing pending compress buffer")
+	}
+
+	b := outBuf.Bytes()
+
+	_, err = crc32Hash.Write(b)
+	if err != nil {
+		return writtenBytes, nil, errors.Wrap(err, "build non-indexed labels hash")
+	}
+
+	n, err = w.Write(b)
+	if err != nil {
+		return writtenBytes, nil, errors.Wrap(err, "write non-indexed labels block")
+	}
+	writtenBytes += n
+	return writtenBytes, crc32Hash.Sum(nil), nil
+}
+
+// CheckpointTo writes all the labels to the writer.
+// It returns back the number of bytes written and a checksum of the data written.
+func (s *symbolizer) CheckpointTo(w io.Writer) (int, []byte, error) {
+	crcHash := crc32HashPool.Get().(hash.Hash32)
+	defer crc32HashPool.Put(crcHash)
+
+	eb := EncodeBufferPool.Get().(*encbuf)
+	defer EncodeBufferPool.Put(eb)
+
+	writtenBytes := 0
+	crcHash.Reset()
+	eb.reset()
+
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	// write the number of labels
+	eb.putUvarint(len(s.labels))
+	n, err := w.Write(eb.get())
+	if err != nil {
+		return 0, nil, err
+	}
+
+	_, err = crcHash.Write(eb.get())
+	if err != nil {
+		return writtenBytes, nil, err
+	}
+
+	eb.reset()
+	writtenBytes += n
+
+	for _, label := range s.labels {
+		// write label length
+		eb.putUvarint(len(label))
+		n, err := w.Write(eb.get())
+		if err != nil {
+			return writtenBytes, nil, err
+		}
+
+		_, err = crcHash.Write(eb.get())
+		if err != nil {
+			return writtenBytes, nil, err
+		}
+
+		eb.reset()
+		writtenBytes += n
+
+		// write the label
+
+		_, err = crcHash.Write(util.YoloBuf(label))
+		if err != nil {
+			return writtenBytes, nil, err
+		}
+
+		n, err = io.WriteString(w, label)
+		if err != nil {
+			return writtenBytes, nil, err
+		}
+		writtenBytes += n
+	}
+
+	return writtenBytes, crcHash.Sum(nil), nil
+}
+
+// symbolizerFromCheckpoint builds symbolizer from the bytes generated during a checkpoint.
+func symbolizerFromCheckpoint(b []byte) *symbolizer {
+	if len(b) == 0 {
+		return newSymbolizer()
+	}
+
+	db := decbuf{b: b}
+	s := symbolizer{
+		symbolsMap: map[string]uint32{},
+	}
+	numLabels := db.uvarint()
+
+	for i := 0; i < numLabels; i++ {
+		label := string(db.bytes(db.uvarint()))
+		s.labels = append(s.labels, label)
+		s.symbolsMap[label] = uint32(i)
+		s.size += len(label)
+	}
+
+	return &s
+}
+
+// symbolizerFromEnc builds symbolizer from the bytes generated during serialization.
+func symbolizerFromEnc(b []byte, pool ReaderPool) (*symbolizer, error) {
+	db := decbuf{b: b}
+	db.uvarint()
+
+	b = db.b
+
+	reader, err := pool.GetReader(bytes.NewBuffer(b))
+	if err != nil {
+		return nil, err
+	}
+	defer pool.PutReader(reader)
+
+	var (
+		readBuf      [10]byte // Enough bytes to store one varint.
+		readBufValid int      // How many bytes are left in readBuf from previous read.
+		s            symbolizer
+		buf          []byte
+	)
+
+	s.compressedSize = len(b)
+
+	for {
```

Comment:
IIUC this loop will read until there is an EOF. Since in the future, we might want to put something after the symbols. I think we should rather store the number of symbols first and then read that many.